File: 11437/main.py
import sys
sys.setrecursionlimit(int(1e9))

input = sys.stdin.readline
n = int(input())

graph = [[] for _ in range(n + 1)]
visit = [False] * (n + 1)
depth = [0] * (n + 1)
parent = [0] * (n + 1)

for _ in range(n - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)


def dfs(x, d):
    visit[x] = True
    depth[x] = d

    for y  in graph[x]:
        if visit[y]:
            continue
        parent[y] = x
        dfs(y, d + 1)


def lca(a, b):
    while depth[a] != depth[b]:
        if depth[a] > depth[b]:
            a = parent[a]
        else:
            b = parent[b]
    while a != b:
        a = parent[a]
        b = parent[b]
    return a


dfs(1, 0)
m = int(input())
for i in range(m):
    a, b = map(int,input().split())
    print(lca(a, b))

File: 11437/cpp/main.cpp
#include <iostream>
#include <bits/stdc++.h>

using namespace std;

vector<int> graph[50001];
int parent[50005];
int d[50005];
int v[5005];

void dfs(int x, int depth){
    v[x] = 1;
    d[x] = depth;
    for (int y : graph[x]){
        if (!v[y]) {
            parent[y] = x;
            dfs(y, depth + 1);
        }
    }
}

int lca(int a, int b){
    while (d[a] != d[b]){
        if (d[a] > d[b]) a = parent[a];
        else b = parent[b];
    }
    while (a != b){
        a = parent[a];
        b = parent[b];
    }
    return a;
}


int main() {
    int n, m;
    cin >> n;

    for (int i = 0; i < n - 1; i ++ ){
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    dfs(1, 0);
    cin >> m;
    int answer[m + 1];
    for (int i = 0; i < m; i ++){
        int a,b;
        cin >> a >> b;
        answer[i] = lca(a, b);
    }
    for (int i = 0; i < m; i++){
        printf("%d\n",answer[i]);
    }
}
